src/parser/ConstructClause.h

```diff
@@ -1,30 +1,62 @@
 //  Copyright 2022, University of Freiburg,
 //  Chair of Algorithms and Data Structures.
 //  Author: Julian Mundhahs (mundhahj@informatik.uni-freiburg.de)
+//  Copyright 2025, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
 
 #pragma once
 
 #include "parser/SelectClause.h"
 #include "parser/data/Types.h"
 
 namespace parsedQuery {
+
+class LazyBindForVariable : public ad_utility::InputRangeFromGet<Variable> {
+ public:
+  LazyBindForVariable(const ad_utility::sparql_types::Triples& triples)
+      : triples_(triples),
+        tripleIt(triples_.begin()),
+        varOrTermIt(tripleIt->begin()) {}
+
+  std::optional<Variable> get() {
+    if (tripleIt == triples_.end()) {
+      return std::nullopt;
+    }
+
+    // use two iterators to find all variables in the triples
+    while (!std::holds_alternative<Variable>(*varOrTermIt)) {
+      ++varOrTermIt;
+      if (varOrTermIt == tripleIt->end()) {
+        ++tripleIt;
+        if (tripleIt == triples_.end()) {
+          return std::nullopt;
+        }
+        varOrTermIt = tripleIt->begin();
+      }
+    }
+
+    const Variable variable = std::get<Variable>(*varOrTermIt);
+    ++varOrTermIt;
+
+    return variable;
+  }
+
+ private:
+  const ad_utility::sparql_types::Triples& triples_;
+  ad_utility::sparql_types::Triples::const_iterator tripleIt;
+  ad_utility::sparql_types::Triple::const_iterator varOrTermIt;
+};
+
 struct ConstructClause : ClauseBase {
   ad_utility::sparql_types::Triples triples_;
 
   ConstructClause() = default;
   explicit ConstructClause(ad_utility::sparql_types::Triples triples)
       : triples_(std::move(triples)) {}
 
-  // Yields all variables that appear in this `ConstructClause`. Variables that
-  // appear multiple times are also yielded multiple times.
-  cppcoro::generator<const Variable> containedVariables() const {
-    for (const auto& triple : triples_) {
-      for (const auto& varOrTerm : triple) {
-        if (auto variable = std::get_if<Variable>(&varOrTerm)) {
-          co_yield *variable;
-        }
-      }
-    }
+  // Lazily yields all variables that appear in this `ConstructClause`.
+  // Variables that appear multiple times are also yielded multiple times.
+  LazyBindForVariable containedVariables() const {
+    return LazyBindForVariable(triples_);
```

Comment:
Can't this just be a concatenation of `ql::ranges/views`, like

return `triples_ | ql::views::join | ql:views::filter(ad_utility::holdsAlternative<Variable>) | ql::views::transform(ad_utility::get<Variable>)`
With return type `auto` (because it is in a header).

Where the two `ad_utility::` functors are defined in `util/TransparentFunctors.h`.

In general a lot of these generators can be reimplemented by stitching together preexisting patterns, some of which I'll implement along the way if missing.